### DISCUSSION TITLE
Welcome to mkdocs_macros_plugin Discussions!

### DISCUSSION THREAD

**Discussion by fralau** [fralau/mkdocs-macros-plugin#56] [Community announcement/discussion]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jmdye** [comment]
Hi. Just came across this plugin for MkDocs. Thanks so much for writing it. I think it might be the answer to a problem I am having. I'm using MkDocs to build a page that displays all of the images in my digital asset library. I have them organized into different directories by asset type (banners, icons, graphics, etc.).

In writing my templates, I basically want to be able to iterate over all of those items and drop them into the template using a simple for loop - something like:
```
{{% for banner in banners_dir %}}
    <img src="{{ banner.url|url }}"/>
{{% endfor %}}
```
What's the best way i could go about accomplishing this?

**Reply by fralau** [reply]
Thanks for using discussions! 

The question would be: what kind of object your banner would be?

Suppose it is just a url (string).

A simple way to implement that, which would require zero Python programming, would be to use the `extra` space in your config file (`mkdocs.yaml`). 

e.g. 

```yaml
extra:
    banners:
        - my.url1.com
        - my.url2.com
        - ...
  ```

Then, in your page:

```
{{% for banner in banners %}}
    <img src="{{ banner }}"/>
{{% endfor %}}
```

That's it!

**Reply by jmdye** [reply]
Thanks for the response. I'm aware of that pattern but because this would be a digital assets library, I'm trying to avoid having to explicitly write the name and path of each asset into a yaml file. That would mean hundreds of lines as I've got several dozen banners, and hundreds of icons and graphics in this library.

My Project is set up like:
```
digital-assets
├── /docs
├── /layouts
│    ├── /assets
│    │    ├── /css
│    │    ├── /img
│    │    │    ├── /banners
│    │    │    │    ├── banner1.jpg
│    │    │    │    ├── banner2.jpg
│    │    │    │    ├── banner3.jpg
│    │    │    │    ├── banner4.jpg
│    │    │    │    ├── banner5.jpg
│    │    │    │    ├── ...
│    │    │    ├── /graphics
│    │    │    ├── /icons
│    │    ├── /js
│    ├── docs.html
│    ├── home.html
│    ├── main.html
│    ├── search.html
├── mkdocs.yml
├── readme.md

```

What I'd like to do if at all possible, is maybe write a macro that could be called within a markdown file or template. 

Something like:
`getBanners`: Returns a dictionary containing an entry for every image file found in the banners folder, where the key is the filename and value is the URL at which the graphic can be accessed over http.

Using that, in a template or md file, I could do something like:
```
{% set banners = {{ get_banners }} %}
{% for banner in banners %}
    {{ banner.url }}
{% endfor %}
```

From there, I could go on later to write a couple of other macros that might get a specific banner, graphic or icon and return it.

**Reply by fralau** [reply]
Oh yes, this would be different.

You would want to scan a directory full of banners, so that it updates dynamically?

The best thing is that you write a _macro_ `get_banners()`, as you surmised.

That macro should go into a `main.yaml` file (see [instructions on how to write a macro](https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/)).

It would look like:

```python
def define_env(env):

    @env.macro
    def get_banners():
          ....
          return .....
```

If you know how to write Python code, scanning the `digital_assets/.../banners` and returning the Banner object you need, should be a piece of cake.

In your markdown file, you might want to write:

```
{% for banner in get_banners() %}
    {{ banner.url }}
{% endfor %}
```

**Reply by jmdye** [reply]
Thanks. I'll take a couple of cracks and post back if I run into an issue

**Reply by jmdye** [reply]
all done. like you said - piece of cake.

**Reply by fralau** [reply]
> all done. like you said - piece of cake.

@jmdye That's a very satisfying statement. Doing things oneself is one thing. Having others doing the same thing and finding it easy, is quite another. 🙂
